### PR TITLE
Direction Finding: HCI_LE_Read_Antenna_Information HCI Command implementation

### PR DESCRIPTION
**Pull request by ppryga-nordic**
This PR is just a start for Bluetooth 5.1 Direction Finding functionality implementation.
I'd like to get input from community about provided changes.

### CODE REVIEW

File: include/bluetooth/hci.h [hunk 1]
```diff
@@ -1335,6 +1335,28 @@ struct bt_hci_cp_le_set_privacy_mode {
...
+#define BT_HCI_LE_NUM_ANTENNAE_MIN              0x1
+#define BT_HCI_LE_NUM_ANTENNAE_MAX              0x4B
+
+#define BT_HCI_LE_MAX_SWITCH_PATTERN_LEN_MIN    0x2
+#define BT_HCI_LE_MAX_SWITCH_PATTERN_LEN_MAX    0x4B
+
+#define BT_HCI_LE_MAX_CTE_LEN_MIN               0x2
+#define BT_HCI_LE_MAX_CTE_LEN_MAX               0x14
+
+#define BT_HCI_OP_LE_READ_ANT_INF               BT_OP(BT_OGF_LE, 0x0058)
```
Comment (cvinayak) [inline include/bluetooth/hci.h:1351]:
```suggestion
#define BT_HCI_OP_LE_READ_ANT_INFO               BT_OP(BT_OGF_LE, 0x0058)
```

File: include/bluetooth/hci.h [hunk 2]
```diff
@@ -1335,6 +1335,28 @@ struct bt_hci_cp_le_set_privacy_mode {
...
+struct bt_hci_rp_le_read_ant_inf {
```
Comment (cvinayak) [inline include/bluetooth/hci.h:1352]:
```suggestion
struct bt_hci_rp_le_read_ant_info {
```

File: include/bluetooth/hci.h [hunk 3]
```diff
@@ -1335,6 +1335,28 @@ struct bt_hci_cp_le_set_privacy_mode {
 	uint8_t         mode;
 } __packed;
 
+#define BT_HCI_LE_1US_SWITCHING_FOR_AOD_TX      BIT(0)
+#define BT_HCI_LE_1US_SAMPLING_FOR_AOD_RX       BIT(1)
+#define BT_HCI_LE_1US_SAMPL_SWITCH_FOR_AOA_RX   BIT(2)
+
+#define BT_HCI_LE_NUM_ANTENNAE_MIN              0x1
+#define BT_HCI_LE_NUM_ANTENNAE_MAX              0x4B
```
Comment (cvinayak) [inline include/bluetooth/hci.h:1343]:
```suggestion
#define BT_HCI_LE_NUM_ANT_MIN              0x1
#define BT_HCI_LE_NUM_ANT_MAX              0x4B
```

File: include/bluetooth/hci.h [hunk 4]
```diff
@@ -1335,6 +1335,28 @@ struct bt_hci_cp_le_set_privacy_mode {
 	uint8_t         mode;
 } __packed;
 
+#define BT_HCI_LE_1US_SWITCHING_FOR_AOD_TX      BIT(0)
+#define BT_HCI_LE_1US_SAMPLING_FOR_AOD_RX       BIT(1)
+#define BT_HCI_LE_1US_SAMPL_SWITCH_FOR_AOA_RX   BIT(2)
```
Comment (cvinayak) [inline include/bluetooth/hci.h:1340]: Use consistent prefix. I have not read the spec., but I am guessing there are options for antennae switching/sampling

File: include/bluetooth/hci.h [hunk 5]
```diff
@@ -1335,6 +1335,28 @@ struct bt_hci_cp_le_set_privacy_mode {
...
+#define BT_HCI_LE_MAX_SWITCH_PATTERN_LEN_MIN    0x2
+#define BT_HCI_LE_MAX_SWITCH_PATTERN_LEN_MAX    0x4B
+
+#define BT_HCI_LE_MAX_CTE_LEN_MIN               0x2
+#define BT_HCI_LE_MAX_CTE_LEN_MAX               0x14
+
+#define BT_HCI_OP_LE_READ_ANT_INF               BT_OP(BT_OGF_LE, 0x0058)
+struct bt_hci_rp_le_read_ant_inf {
+	uint8_t status;
+	uint8_t switch_sampl_rates;
```
Comment (cvinayak) [inline include/bluetooth/hci.h:1354]:
```suggestion
	uint8_t switch_sample_rates;
```

File: subsys/bluetooth/host/CMakeLists.txt [hunk 6]
```diff
@@ -57,6 +57,12 @@ if(CONFIG_BT_HCI_HOST)
         )
     endif()
   endif()
+
+  if(CONFIG_BT_DF)
+    zephyr_library_sources(
+      df.c
```
Comment (cvinayak) [inline subsys/bluetooth/host/CMakeLists.txt:63]:
```suggestion
      direction.c
```

File: subsys/bluetooth/host/Kconfig [hunk 7]
```diff
@@ -598,6 +598,14 @@ config BT_ID_MAX
 	  Maximum number of supported local identity addresses. For most
 	  products this is safe to leave as the default value (1).
 
+config BT_DF
```
Comment (cvinayak) [inline subsys/bluetooth/host/Kconfig:601]:
```suggestion
config BT_DIRECTION_FINDING
```

File: subsys/bluetooth/host/Kconfig [hunk 8]
```diff
@@ -598,6 +598,14 @@ config BT_ID_MAX
...
+	bool "Enable Direction Finding support"
```
Comment (cvinayak) [inline subsys/bluetooth/host/Kconfig:602]:
```suggestion
	bool "Enable Direction Finding support [EXPERIMENTAL]"
```

File: subsys/bluetooth/host/df.h [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+#include <sys/util.h>
+
+enum df_opt_swich_sampl_rates {
+	/* Support Angle of Departure 1us switching TX */
+	DF_AOD_1US_SWITCH_TX_SUPP = BIT(0),
+	/* Support Angle of departure 1us sampling RX */
+	DF_AOD_1US_SAMPL_RX_SUPP = BIT(1),
+	/* Support Angle of Arrival 1us switching and sampling RX */
+	DF_AOA_1US_SWITCH_SAMPL_RX_SUPP  = BIT(2),
+};
```
Comment (cvinayak) [inline subsys/bluetooth/host/df.h:19]: Reuse HCI defines

File: subsys/bluetooth/host/df.h [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+/* @brief Antenna information for LE Direction Finding */
+struct bt_le_df_ant_inf {
+	uint8_t switch_sampl_rates;
+	uint8_t num_ant;
+	uint8_t max_switch_pattern_len;
+	uint8_t max_CTE_len;
+};
```
Comment (cvinayak) [inline subsys/bluetooth/host/df.h:27]: Why not reuse HCI info structure?

**Review by cvinayak — CHANGES_REQUESTED** [review]

**Comment by ppryga** [comment]
I've got a question related with processing with all other HCI commands. 
I wanted to implement all other DF HCI commands in Host in a similar way as I did with HCI_LE_Read_Antenna_Information. 
All commands would be implemented in separate commits and provided as single PR.
Is that acceptable approach?

**Comment by cvinayak** [comment]
> All commands would be implemented in separate commits and provided as single PR.
> Is that acceptable approach?

Yes, this is ok.


Please correct commit message title:
tests: bluetoot: Add test solution for DF HCI commands

**Comment by ppryga** [comment]
> >All commands would be implemented in separate commits and provided as single PR.
> >Is that acceptable approach?
> 
> Yes, this is ok.
Great!

I will fix commit message when applying requested changes.

File: include/bluetooth/hci.h (hunk 4 above)
Comment (ppryga) [reply include/bluetooth/hci.h:1340]: Actually this one is inconsistent... Those defines are related with non-mandatory modes for AoD and AoA:

From BT Spec Vol 4: 7.8.87
Supported_Switching_Sampling_Rates:
- bit nr: 0 "1 μs switching supported for AoD transmission"
- bit nr: 1 "1 μs sampling supported for AoD reception"
- bit nr: 2 "1 μs switching and sampling supported for AoA reception"
- bit nr: 3-7 "Reserved for future use"

2μs switching and sampling is mandatory.

File: subsys/bluetooth/controller/Kconfig [hunk 11]
```diff
@@ -457,6 +457,11 @@ config BT_CTLR_ASSERT_HANDLER
 	  and will be invoked whenever the controller code encounters
 	  an unrecoverable error.
 
+config BT_CTRL_DF
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/Kconfig:460]: Typo: `CTRL -> CTLR`

Typically, a DF application only consist of parts of the entire DF specification. Therefore, the specification allows
to only qualify parts of it. See Core_v5.2, Vol 6, Section 4.6.
A suggested split of configurations is provided below.
* `BT_CTLR_DF_CONN_REQ`
* `BT_CTLR_DF_CONN_RSP`
* `BT_CTLR_DF_BROADCAST_RX`
* `BT_CTLR_DF_BROADCAST_TX`
* `BT_CTLR_DF_TX_AOA`
* `BT_CTLR_DF_RX`
* `BT_CTLR_DF_RX_AOD`

That will allow more fine grained control of which DF features that are included in the controller. 

To hide configurations that are not supported by a given controller, these configurations should have their corresponding `_SUPPORT` configurations which they depend upon. See for example `BT_CTLR_SLAVE_FEAT_REQ`.

**Review by rugeGerritsen — CHANGES_REQUESTED** [review]
Applications should be able to only include parts of the entire DF implementation. See suggested split of configurations.

File: subsys/bluetooth/host/df.h (hunk 10 above)
Comment (ppryga) [reply subsys/bluetooth/host/df.h:27]: I thought that types declared in hci.h are used in HCI related source code.

The aim of this declaration is to provide a type to use it in implementation of higher level code. E.g. in future if DF related services or profiles are defined. Right now this part of code is used only by test code and may use HCI info structure as well. Question here is about current approach for such situations. Is it expected to have a separation of types between "layers" or reuse types as much as possible?

File: include/bluetooth/hci.h (hunk 1 above)
Comment (ppryga) [reply include/bluetooth/hci.h:1351]: Corrected

File: include/bluetooth/hci.h (hunk 2 above)
Comment (ppryga) [reply include/bluetooth/hci.h:1352]: Corrected

File: include/bluetooth/hci.h (hunk 3 above)
Comment (ppryga) [reply include/bluetooth/hci.h:1343]: Corrected

File: include/bluetooth/hci.h (hunk 4 above)
Comment (ppryga) [reply include/bluetooth/hci.h:1340]: What do you think if I use following names?

#define BT_HCI_LE_1US_AOD_TX	BIT(0)
#define BT_HCI_LE_1US_AOD_RX	BIT(1)
#define BT_HCI_LE_1US_AOA_RX	BIT(2)

They are more consistent. Also AOD_TX is implicitly related with switching and AOD_RX is implicitly related with sampling so there is no need to put it in names.

Comment (ppryga) [reply include/bluetooth/hci.h:1340]: Changed to last proposed by me version

File: include/bluetooth/hci.h (hunk 5 above)
Comment (ppryga) [reply include/bluetooth/hci.h:1354]: Corrected

File: subsys/bluetooth/host/CMakeLists.txt (hunk 6 above)
Comment (ppryga) [reply subsys/bluetooth/host/CMakeLists.txt:63]: Corrected

File: subsys/bluetooth/host/Kconfig (hunk 7 above)
Comment (ppryga) [reply subsys/bluetooth/host/Kconfig:601]: Corrected

File: subsys/bluetooth/host/Kconfig (hunk 8 above)
Comment (ppryga) [reply subsys/bluetooth/host/Kconfig:602]: Corrected

File: subsys/bluetooth/host/df.h (hunk 9 above)
Comment (ppryga) [reply subsys/bluetooth/host/df.h:19]: See my comment in following question

File: subsys/bluetooth/controller/Kconfig (hunk 11 above)
Comment (ppryga) [reply subsys/bluetooth/controller/Kconfig:460]: Configuration will be split into fine grained items that follow Core_v5.2, Vol 6, Section 4.6. DF configuration will be put in separate KConfig file and included in main controller KConfig.

**Comment by joerchan** [comment]
Looks like it is a duplicating commits in #29603?
I've posted my comments there.

**Comment by ppryga** [comment]
I close this PR since the same commits are delivered in #29603. Please add further comments there.

**Pull request closed without merging**